parquet/read.go

```diff
@@ -59,120 +54,174 @@ func (*Client) Read(f ReaderAtSeeker, table *schema.Table, _ string, res chan<-
-			}
-			cols[c] = sb.NewArray()
-		case arrow.TypeEqual(arrowSchema.Field(c).Type, types.NewJSONType()):
-			arr := col.(*array.String)
-			b, err := arr.MarshalJSON()
-			if err != nil {
-				return nil, fmt.Errorf("failed to marshal col %v: %w", rec.ColumnName(c), err)
-			}
-			a := make([]any, arr.Len())
-			err = json.Unmarshal(b, &a)
-			if err != nil {
-				return nil, fmt.Errorf("failed to unmarshal col %v: %w", rec.ColumnName(c), err)
-			}
-			sb := types.NewJSONBuilder(array.NewExtensionBuilder(memory.DefaultAllocator, types.NewJSONType()))
-			for _, v := range a {
-				if v == nil {
-					sb.AppendNull()
-					continue
-				}
-				var v2 any
-				err = json.Unmarshal([]byte(v.(string)), &v2)
-				if err != nil {
-					return nil, fmt.Errorf("failed to unmarshal col %v: %w", rec.ColumnName(c), err)
-				}
-				sb.Append(v2)
-			}
-			cols[c] = sb.NewArray()
-		case arrow.TypeEqual(arrowSchema.Field(c).Type, types.NewMacType()):
-			arr := col.(*array.String)
-			b, err := arr.MarshalJSON()
-			if err != nil {
-				return nil, fmt.Errorf("failed to marshal col %v: %w", rec.ColumnName(c), err)
-			}
-			sb := types.NewMacBuilder(array.NewExtensionBuilder(memory.DefaultAllocator, types.NewMacType()))
-			err = sb.UnmarshalJSON(b)
-			if err != nil {
-				return nil, fmt.Errorf("failed to unmarshal col %v: %w", rec.ColumnName(c), err)
+	for i := 0; i < int(rec.NumCols()); i++ {
+		cols[i] = reverseTransformArray(sc.Field(i), rec.Column(i))
+	}
+	return array.NewRecord(sc, cols, -1)
+}
+
+func reverseTransformArray(f arrow.Field, col arrow.Array) arrow.Array {
+	dt := f.Type
+	switch {
+	case arrow.TypeEqual(dt, types.ExtensionTypes.UUID):
+		return reverseTransformUUID(col.(*array.String))
+	case arrow.TypeEqual(dt, types.ExtensionTypes.Inet):
+		return reverseTransformInet(col.(*array.String))
+	case arrow.TypeEqual(dt, types.ExtensionTypes.MAC):
+		return reverseTransformMAC(col.(*array.String))
+	case arrow.TypeEqual(dt, types.ExtensionTypes.JSON):
+		return reverseTransformJSON(col.(*array.String))
+	case arrow.TypeEqual(col.DataType(), arrow.FixedWidthTypes.Timestamp_ms):
+		return reverseTransformTimestamp(dt.(*arrow.TimestampType), col.(*array.Timestamp))
+	case dt.ID() == arrow.STRUCT:
+		return reverseTransformStruct(dt.(*arrow.StructType), col.(*array.String))
+	case arrow.IsListLike(dt.ID()) && dt.(*arrow.ListType).Elem().ID() == arrow.EXTENSION:
+		child := reverseTransformArray(
+			arrow.Field{
+				Type: dt.(*arrow.ListType).Elem(),
+				Name: "list<ext:" + dt.(*arrow.ListType).Elem().Name() + ">",
+			},
+			col.(*array.List).ListValues(),
+		)
+		fmt.Println("counts", col.Len(), child.Len(), col.NullN(), child.NullN()) //  1 3 0 1
+
+		return array.NewExtensionData(array.NewData(dt.(*arrow.ListType).Elem(), child.Len(), child.Data().Buffers(), []arrow.ArrayData{child.Data()}, child.NullN(), child.Data().Offset()))
+
+	case arrow.IsListLike(dt.ID()):
```

Comment:
This will dramatically fail, as it'll basically flatten the list underneath.
Instead, you need to use [`array.ListLike`](https://pkg.go.dev/github.com/apache/arrow/go/v13/arrow/array#ListLike) & [`array.NewSlice`](https://pkg.go.dev/github.com/apache/arrow/go/v13/arrow/array#NewSlice) like this:
```go
list := col.(array.ListLike)
builder := array.NewBuilder(memory.DefaultAllocator, dt).(array.ListLikeBuilder)
valBuilder := builder.ValueBuilder()
for i := 0; i < list.Len(); i++ {
	if list.IsNull(i) {
		builder.AppendNull()
		continue
	}
	builder.Append(true)
	start, end := list.ValueOffsets(i)
	fillInArr(valBuilder, array.NewSlice(list.ListValues(), start, end))
}
```